### ISSUE TITLE
FSDP consumes the same amount of memory as DPP，why？

### ISSUE THREAD

**Issue by Alex-Songs** [facebookresearch/fairscale#1050]
Hello.
I am using FSDP to replace DDP, but I find that the memory consumption is the same. Is it because FSDP only supports nn.Sequential model （my model is not a nn.Sequential model ）?

**Comment by min-xu-ai** [comment]
Can you provide more details? There is a number of reasons:

1. maybe you didn't measure the memory correctly. Pytorch does not research memory from CUDA in an eager fashion so you may not see the memory saving.
2. your model may be small and main memory consumption might be from activation tensors saved for the backward. In that case, you won't see memory saving from sharding
3. you are only sharding among 2 GPUs, which doesn't show the saving significantly for your model size
4. you didn't use nested wrapping of FSDP to make model memory sharding happen on all layers that are not being actively computing.

There are other reasons I may not be thinking about. Overall, FSDP will save significant amount of memory when model is big and sharding is done over a large number of GPUs

**Comment by Alex-Songs** [comment]
@min-xu-ai 
Thanks，
I replaced the DPP in the code with the FSDP by the following，and uses the AMP model of pytorch.：
<img width="777" alt="image" src="https://user-images.githubusercontent.com/67909291/182316871-f67245a1-584c-4eaf-aee3-cf533571ec58.png">

Testing on two GPUs, we found that FSDP consumes more memory: 
DDP：
![image](https://user-images.githubusercontent.com/67909291/182317208-4f5d508b-f5b5-4197-95f6-f5b4abd22cc3.png)
FSDP：
<img width="544" alt="image" src="https://user-images.githubusercontent.com/67909291/182317276-93920738-b1b8-4613-91a8-c414e467c4de.png">

**Comment by min-xu-ai** [comment]
Try reshard after forward = true. Also, make sure you use more gpus and activation is not the main memory consumer

**Comment by Alex-Songs** [comment]
@min-xu-ai 
Thanks。
I suspected that the model was too small, causing the FSDP to wrap the entire model, so I set min_num_params to 20000 and the model was able to have multiple FSDP units, but the training got stuck at：
![image](https://user-images.githubusercontent.com/67909291/182394435-65650c62-bd72-4194-b0f3-ff2139d0b7aa.png)

**Comment by min-xu-ai** [comment]
That’s interesting. Do you have a small reproducible case?

If I have to guess it might be related to different rank compute on different code paths. An example is layerdrop, I.e. dropping layers on forward pass randomly

**Comment by Alex-Songs** [comment]
@min-xu-ai 
Thanks，the reason is the layer drop op.
I have another question. When I tried reshard after forward = true, the following error occurred: 
WFPB: incorrect saved_grad_shard device CPU vs CUDA :0

**Comment by min-xu-ai** [comment]
Are you using no_sync() context and doing gradient accumulation? Or are you doing cpu_offloading? You might need to disable one of them.

**Comment by Alex-Songs** [comment]
@min-xu-ai 
I have replaced no_sync() with A and turned off cpu_offloading, but I still have this error.
<img width="402" alt="image" src="https://user-images.githubusercontent.com/67909291/183480252-1599d933-33b2-4082-aeef-4b31eecaf28f.png">
<img width="560" alt="image" src="https://user-images.githubusercontent.com/67909291/183480430-54bbde00-b00e-4b22-a5f9-861175635cee.png">
<img width="987" alt="image" src="https://user-images.githubusercontent.com/67909291/183480520-d1ac277b-7c55-4c4b-912b-a26980f5e2ee.png">

**Comment by Alex-Songs** [comment]
@min-xu-ai 
In addition, when I use cpu_offload I get the following error：
<img width="1044" alt="image" src="https://user-images.githubusercontent.com/67909291/183486851-9fdd48b3-f31c-4755-aa1b-fe0c074cbf2b.png">

**Comment by min-xu-ai** [comment]
If you have a small reproducible test case (a single script that I can run to reproduce the issue), I can try to look into it.

**Comment by min-xu-ai** [comment]
Actually, `no_sync` with FSDP is a bit different from that of DDP. With `no_sync`, you actually won't get memory benefit because it would hold on to full params and full gradients, which makes it the same as DDP. This has the benefit of reduced network communication.

However, if you want to save memory in the expense of the more network communication, you can do gradient accumulation directly with FSDP, without no_sync, i.e.

```
for i in range(n):
  loss = model(inputs[i])
  loss.backward()

optim.step()
```

FSDP will internally do gradient accumulation above and you can do a step() after your accumulation steps are done.

See source here:  https://github.com/facebookresearch/fairscale/blob/6f03e41542e7ccd78d49894860b7301066cca857/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1103

Will close this and please reopen if you have additional questions.
